File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-valid-physical-material-properties.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticObjectPropertyValue } from "../../utils/get-static-object-property-value.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  PHYSICAL_MATERIAL_IOR_PROPERTY_NAMES,
  PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES,
} from "./constants.js";
import { getApiReferenceProvenance } from "./utils/get-api-reference-provenance.js";
import { getInvalidPhysicalMaterialProperty } from "./utils/get-invalid-physical-material-property.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { getThreePropertyAssignment } from "./utils/get-three-property-assignment.js";
import { isThreeModuleSource } from "./utils/is-three-module-source.js";

const PHYSICAL_MATERIAL_PROPERTY_NAMES: ReadonlySet<string> = new Set([
  ...PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES,
  ...PHYSICAL_MATERIAL_IOR_PROPERTY_NAMES,
]);

const reportInvalidProperty = (
  propertyName: string,
  expression: EsTreeNode,
  context: RuleContext,
): void => {
  const value = getStaticNumber(expression, context.scopes);
  if (value === null) return;
  const invalidProperty = getInvalidPhysicalMaterialProperty(propertyName, value, expression);
  if (!invalidProperty) return;
  context.report({
    node: expression,
    message: `${propertyName} is ${String(value)}, but MeshPhysicalMaterial requires ${propertyName} in [${String(invalidProperty.minimum)}, ${String(invalidProperty.maximum)}]`,
  });
};

export const threeValidPhysicalMaterialProperties = defineRule({
  id: "three-valid-physical-material-properties",
  title: "Invalid Three.js physical material property",
  category: "Correctness",
  severity: "warn",
  recommendation:
    "Keep MeshPhysicalMaterial layer strengths and indices of refraction in their documented ranges",
  create: (context: RuleContext) => ({
    NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
      const provenance = getApiReferenceProvenance(node.callee, context.scopes);
      const parameters = node.arguments[0];
      if (
        provenance?.apiName !== "MeshPhysicalMaterial" ||
        !isThreeModuleSource(provenance.moduleSource) ||
        !parameters
      ) {
        return;
      }
      for (const propertyName of PHYSICAL_MATERIAL_PROPERTY_NAMES) {
        const expression = getStaticObjectPropertyValue(parameters, propertyName);
        if (expression) reportInvalidProperty(propertyName, expression, context);
      }
    },
    AssignmentExpression(node: EsTreeNodeOfType<"AssignmentExpression">) {
      const assignment = getThreePropertyAssignment(node, context);
      if (
        assignment?.constructorName !== "MeshPhysicalMaterial" ||
        !PHYSICAL_MATERIAL_PROPERTY_NAMES.has(assignment.propertyName)
      ) {
        return;
      }
      reportInvalidProperty(assignment.propertyName, assignment.value, context);
    },
  }),
});

File: native/oxlint/rules/three-valid-physical-material-properties.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;
use oxc_syntax::operator::AssignmentOperator;

use crate::{context::LintContext, rule::Rule, AstNode};

const PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES: [&str; 9] = [
    "anisotropy",
    "clearcoat",
    "clearcoatRoughness",
    "iridescence",
    "reflectivity",
    "sheen",
    "sheenRoughness",
    "specularIntensity",
    "transmission",
];
const PHYSICAL_MATERIAL_IOR_PROPERTY_NAMES: [&str; 2] = ["ior", "iridescenceIOR"];
const MINIMUM_PHYSICAL_MATERIAL_FACTOR: f64 = 0.0;
const MAXIMUM_PHYSICAL_MATERIAL_FACTOR: f64 = 1.0;
const MINIMUM_PHYSICAL_MATERIAL_IOR: f64 = 1.0;
const MAXIMUM_PHYSICAL_MATERIAL_IOR: f64 = 2.333;

#[derive(Debug, Default, Clone)]
pub struct ThreeValidPhysicalMaterialProperties;

declare_oxc_lint!(
    /// Require documented Three.js physical material property ranges.
    ThreeValidPhysicalMaterialProperties,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Validate Three.js physical material properties.",
);

impl Rule for ThreeValidPhysicalMaterialProperties {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        match node.kind() {
            AstKind::NewExpression(new_expression)
                if three_module_api_path_matches(
                    &new_expression.callee,
                    &["MeshPhysicalMaterial"],
                    ctx,
                ) =>
            {
                let Some(parameters) = new_expression
                    .arguments
                    .first()
                    .and_then(oxc_ast::ast::Argument::as_expression)
                else {
                    return;
                };
                for property_name in PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES
                    .iter()
                    .chain(PHYSICAL_MATERIAL_IOR_PROPERTY_NAMES.iter())
                {
                    if let Some(property_value) =
                        get_static_object_property_value(parameters, property_name)
                    {
                        report_invalid_physical_material_property(
                            property_name,
                            property_value,
                            ctx,
                        );
                    }
                }
            }
            AstKind::AssignmentExpression(assignment)
                if assignment.operator == AssignmentOperator::Assign =>
            {
                let Some(member_expression) = assignment.left.as_member_expression() else {
                    return;
                };
                let Some(property_name) = member_expression.static_property_name() else {
                    return;
                };
                if !is_physical_material_property(property_name)
                    || three_constructor_name(
                        member_expression.object(),
                        &["MeshPhysicalMaterial"],
                        ctx,
                    )
                    .is_none()
                {
                    return;
                }
                report_invalid_physical_material_property(property_name, &assignment.right, ctx);
            }
            _ => {}
        }
    }
}

fn is_physical_material_property(property_name: &str) -> bool {
    PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES.contains(&property_name)
        || PHYSICAL_MATERIAL_IOR_PROPERTY_NAMES.contains(&property_name)
}

fn report_invalid_physical_material_property<'a>(
    property_name: &str,
    expression: &oxc_ast::ast::Expression<'a>,
    ctx: &LintContext<'a>,
) {
    let Some(value) = resolve_static_number(expression, ctx) else {
        return;
    };
    let (minimum, maximum) = if PHYSICAL_MATERIAL_NORMALIZED_PROPERTY_NAMES.contains(&property_name)
    {
        (
            MINIMUM_PHYSICAL_MATERIAL_FACTOR,
            MAXIMUM_PHYSICAL_MATERIAL_FACTOR,
        )
    } else {
        (MINIMUM_PHYSICAL_MATERIAL_IOR, MAXIMUM_PHYSICAL_MATERIAL_IOR)
    };
    if (minimum..=maximum).contains(&value) {
        return;
    }
    ctx.diagnostic(
        OxcDiagnostic::warn(format!(
            "{property_name} is {value}, but MeshPhysicalMaterial requires {property_name} in [{minimum}, {maximum}]"
        ))
        .with_label(expression.span()),
    );
}
